pkg/pillar/hypervisor/hypervisor.go

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2017-2020 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package hypervisor
+
+import (
+	"fmt"
+	"os"
+)
+
+// Hypervisor provides methods for manipulating domains on the host
+type Hypervisor interface {
+	Name() string
+
+	Create(string, string) (int, error)
+
+	Start(string, int) error
+	Tune(string, int, int) error
+	Stop(string, int, bool) error
+	Delete(string, int) error
+	Info(string, int) error
+	LookupByName(string, int) (int, error)
+
+	IsDeviceModelAlive(int) bool
+
+	PCIReserve(string) error
+	PCIRelease(string) error
+}
+
+type hypervisorDesc struct {
+	constructor func() Hypervisor
+	dom0handle  string
+}
+
+var knownHypervisors = map[string]hypervisorDesc{
+	"xen":  {constructor: newXen, dom0handle: "/proc/xen"},
+	"kvm":  {constructor: newKvm, dom0handle: "/dev/kvm"},
+	"acrn": {constructor: newAcrn, dom0handle: "/dev/acrn"},
+	"null": {constructor: newNull, dom0handle: ""},
+}
+
+// GetHypervisor returns a particular hypervisor implementation
+func GetHypervisor(hint string) (Hypervisor, error) {
+	if _, found := knownHypervisors[hint]; !found {
+		return nil, fmt.Errorf("Unknown hypervisor %s", hint)
+	} else {
+		return knownHypervisors[hint].constructor(), nil
```

Comment:
Do you want this code to check the dom0handle, or have the constructor do that? In the latter case I think you should add that check to the acorn, kvm (and xen) constructors so that GetAvailableHypervisors and GetHyperVisor are more constent.

Reply:
This is by design -- we trust the calling code to know what it is doing and not second-guessing it. It may be helpful in variety of different situations (including testing, separation of containers, etc.) to enable a hypervisor that does't seem to be "present" in the current linux kernel.